src/Servers/Kestrel/Core/src/Internal/Http3/Helpers/VariableIntHelper.cs

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Diagnostics;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http3
+{
+    public static class VariableIntHelper
+    {
+        public static long GetVariableIntErrorIfNotAllowedValue(in ReadOnlySequence<byte> buffer, out SequencePosition consumed, out SequencePosition examined)
+        {
+            // 0x1F * N + 0x21 isn't allowed by spec, return -1 if that's the case.
+            // https://quicwg.org/base-drafts/draft-ietf-quic-http.html#frame-reserved
+            var longLength = GetVariableIntFromReadOnlySequence(buffer, out consumed, out examined);
+            if ((longLength - 0x21) % 0x1F == 0)
+            {
+                return -1;
+            }
+
+            return longLength;
+        }
+
+        public static long GetVariableIntFromReadOnlySequence(in ReadOnlySequence<byte> buffer, out SequencePosition consumed, out SequencePosition examined)
+        {
+            consumed = buffer.Start;
+            examined = buffer.End;
+
+            if (buffer.Length == 0)
+            {
+                return -1;
+            }
+
+            var ros = buffer.Slice(0, 1);
```

Comment:
As an efficiency thing, I would slice 8 bytes up front, convert that to a Span, then work on it directly.

Reply:
The ReadOnlySequence can be a single byte, right? Hence I'm checking the length like this.

Reply:
It's inefficient to use the `ReadOnlySequence` for a small number of bytes. Since you know the max length is 8 bytes, just slice that much (`Slice(0, Min(Length, 8))`, make a Span and use the span for these operations.

Reply:
Smart idea to do Math.Min();